src/couch_stats/src/csrt_util.erl

```diff
@@ -0,0 +1,485 @@
+    Rctx#rctx{docs_written = Val};
+map_to_rctx_field(js_filter, Val, Rctx) ->
+    Rctx#rctx{js_filter = Val};
+map_to_rctx_field(js_filtered_docs, Val, Rctx) ->
+    Rctx#rctx{js_filtered_docs = Val};
+map_to_rctx_field(rows_read, Val, Rctx) ->
+    Rctx#rctx{rows_read = Val};
+map_to_rctx_field(type, Val, Rctx) ->
+    Rctx#rctx{type = Val};
+map_to_rctx_field(get_kp_node, Val, Rctx) ->
+    Rctx#rctx{get_kp_node = Val};
+map_to_rctx_field(get_kv_node, Val, Rctx) ->
+    Rctx#rctx{get_kv_node = Val};
+map_to_rctx_field(write_kp_node, Val, Rctx) ->
+    Rctx#rctx{write_kp_node = Val};
+map_to_rctx_field(write_kv_node, Val, Rctx) ->
+    Rctx#rctx{write_kv_node = Val};
+map_to_rctx_field(changes_returned, Val, Rctx) ->
+    Rctx#rctx{changes_returned = Val};
+map_to_rctx_field(changes_processed, Val, Rctx) ->
+    Rctx#rctx{changes_processed = Val};
+map_to_rctx_field(ioq_calls, Val, Rctx) ->
+    Rctx#rctx{ioq_calls = Val}.
+
+-spec field(Field :: rctx_field(), Rctx :: rctx()) -> any().
+field(updated_at, #rctx{updated_at = Val}) ->
+    Val;
+field(started_at, #rctx{started_at = Val}) ->
+    Val;
+field(pid_ref, #rctx{pid_ref = Val}) ->
+    Val;
+field(nonce, #rctx{nonce = Val}) ->
+    Val;
+field(dbname, #rctx{dbname = Val}) ->
+    Val;
+field(username, #rctx{username = Val}) ->
+    Val;
+field(db_open, #rctx{db_open = Val}) ->
+    Val;
+field(docs_read, #rctx{docs_read = Val}) ->
+    Val;
+field(docs_written, #rctx{docs_written = Val}) ->
+    Val;
+field(js_filter, #rctx{js_filter = Val}) ->
+    Val;
+field(js_filtered_docs, #rctx{js_filtered_docs = Val}) ->
+    Val;
+field(rows_read, #rctx{rows_read = Val}) ->
+    Val;
+field(type, #rctx{type = Val}) ->
+    Val;
+field(get_kp_node, #rctx{get_kp_node = Val}) ->
+    Val;
+field(get_kv_node, #rctx{get_kv_node = Val}) ->
+    Val;
+field(changes_returned, #rctx{changes_returned = Val}) ->
+    Val;
+field(changes_processed, #rctx{changes_processed = Val}) ->
+    Val;
+field(ioq_calls, #rctx{ioq_calls = Val}) ->
+    Val.
+
+add_delta({A}, Delta) -> {A, Delta};
+add_delta({A, B}, Delta) -> {A, B, Delta};
+add_delta({A, B, C}, Delta) -> {A, B, C, Delta};
+add_delta({A, B, C, D}, Delta) -> {A, B, C, D, Delta};
+add_delta({A, B, C, D, E}, Delta) -> {A, B, C, D, E, Delta};
+add_delta({A, B, C, D, E, F}, Delta) -> {A, B, C, D, E, F, Delta};
+add_delta({A, B, C, D, E, F, G}, Delta) -> {A, B, C, D, E, F, G, Delta};
+add_delta(T, _Delta) -> T.
+
+extract_delta({A, {delta, Delta}}) -> {{A}, Delta};
+extract_delta({A, B, {delta, Delta}}) -> {{A, B}, Delta};
+extract_delta({A, B, C, {delta, Delta}}) -> {{A, B, C}, Delta};
+extract_delta({A, B, C, D, {delta, Delta}}) -> {{A, B, C, D}, Delta};
+extract_delta({A, B, C, D, E, {delta, Delta}}) -> {{A, B, C, D, E}, Delta};
+extract_delta({A, B, C, D, E, F, {delta, Delta}}) -> {{A, B, C, D, E, F}, Delta};
+extract_delta({A, B, C, D, E, F, G, {delta, Delta}}) -> {{A, B, C, D, E, F, G}, Delta};
+extract_delta(T) -> {T, undefined}.
```

Comment:
This looks a bit unusual, maybe we can add an extra map to the messages which can have an optional `delta`: val KV?